### DISCUSSION TITLE
Improve code generation performance

### DISCUSSION THREAD

**Discussion by overbit** [amplication/amplication#6747] [Ideas]
- Improve `build-manager` by leveraging zip copy to prepare dsg-jobs output to artefact folder
- Add cache on dsg generation steps
  - cache based on on hash of dsg-spec 
  - cache based on hash of entity + pluginInstallation spec 
  - ..

**Comment by abrl91** [comment]
- separate between Admin and Server and add the ability to choose whether to build both or one of them
- choose the service/services to build, in case there is more than one service in the project

**Comment by mulygottlieb** [comment]
- parallel the generation of different entities
- parallel the generation of files within each entity

In terms of paralleling things
- Use promise.all() in order not to wait promise after promise
- Use Node.js worker threads (thread pool) e.g. divide the entities to X groups and each thread generates the files for its group
- Same as above but with subprocess (create small executables that generate per entity, etc. and run them from an orchestrator "main" DSG)

**Comment by mulygottlieb** [comment]
In order to test which areas in the DSG are good candidates for improvement, it is advisable to create a very large build and observe the timestamps on the DSG logs in order to identify steps that take a long time and are, therefore, good candidates for improvement attempts.
The easiest way to get to a long build is to import a very large number of entities. We have some ready examples for such files; please ask me if you need such a schema file you can import to generate 200+ entities and observe a long DSG build.

**Comment by mulygottlieb** [comment]
It is advisable to assist in the following "event tree" to remind us of areas we can consider parallelizing:
![image](https://github.com/amplication/amplication/assets/1912933/b8586cf9-d0db-4fef-9ce9-c66da29b1665)

**Comment by mulygottlieb** [comment]
- Split AdminUI and Server creation - Use Redis distributed cache for the Build Manager to sync the responses of both DSG jobs
- Once we add Redis to our Architecture, we can use it in the DSG as well - either to cache some of the results (generated modules) and/or to cache the DSG Context instead of an in-memory Singelton
